### PR TITLE
Sync thread adjustments with wasm-integration

### PR DESCRIPTION
**Pull request by jerinphilip**
Opened at the suggestion of @abhi-agg. 

This PR attempts to synchronize parallel edits pertaining to multithreading/single-threading to provide a cleaner solution for non-multithreaded launches. WASM dev needs a successful build without thread constructs for initial implementation. 

There are option changes in this PR which break wasm-integration.  

- [x] `max-input-sentence-tokens` -> `max-length-break`
- [x] `max-input-tokens` -> `mini-batch-words`. 

WASM is identified to be not using @ugermann's updated ssplit-cpp (with PCRE2) for the branch. 
- [x] ssplit is broken.

### CODE REVIEW

File: CMakeLists.txt
Comment (jerinphilip) [inline CMakeLists.txt:30]: What is `COMPILE_WITH_PTHREADS`, `COMPILE_THREAD_VARIANT` seems to be what's activating the no threads source through `ifdef`s. This is a bit confusing with `WITH_PTHREADS` being the `ifdef` check.

File: src/translator/CMakeLists.txt
Comment (jerinphilip) [inline src/translator/CMakeLists.txt:31]: `WITH_PTHREADS` fro `COMPILE_THREAD_VARIANT` and not `COMPILE_WITH_PTHREADS`.

File: src/translator/TranslationModel.cpp
Comment (jerinphilip) [inline src/translator/TranslationModel.cpp:19]: Merge artifact. TODO(@jerinphilip)

File: src/translator/batch_translator.h
Comment (jerinphilip) [inline src/translator/batch_translator.h:49]: Merge artifact TODO(@jerinphilip)

File: wasm/test_page/bergamot.html
Comment (jerinphilip) [inline wasm/test_page/bergamot.html:85]: Options have changed.

File: src/translator/batch_translator.h
Comment (jerinphilip) [reply src/translator/batch_translator.h]: https://github.com/browsermt/bergamot-translator/pull/30/commits/d72343567c978d9a8605193f7027c23edff728b3

**Comment by jerinphilip** [comment]
@abhi-agg hacked together a threadless implementation which edits concurrent code I'm actively working on. I've taken it on myself to provide an improved thread free compilation path (my modifications switches out concurrent calls and structures instead of turning off locks and mutexes and creating dummy Semaphores). wasm-integration and our edits have since diverged. 

Following a meeting, I ran a dry run at what a merge would look like and what issues can potentially arise from merging wasm-integration -> integration and then integration->main.

The current state of the PR and associated branch are: 
1. This uses browsermt/marian-dev@wasm
2. Build fails at ssplit-cpp. It is not worthwhile to spend more hours putting PCRECPP in, so I leave it to whoever's involved to change this. You have an option of putting ifdef guards in a particular place, so pcrecpp compiles I can point you to where, if somehow the decision is still to keep PCRECPP for WASM related reasons.

On slack, @abhi-agg has asked what are issues in a merge.

From my side, the following are required.

- [ ] Mozilla should provide one easy switch to get pre-wasm, all targets compilation. My compilation looks like the following on this branch, which is not acceptable. Whatever CMakeFile.txt edits are present should be reviewed. I'm okay with existing marian, but I spent a lot of time trying to find these flags and turning them off to get compile to work. Also I expect the code I'm developing to be functioning with all of marian, not a select bunch of targets.

```
$ cmake .. -DCOMPILE_CUDA=off -DCMAKE_BUILD_TYPE=Release \
     -DCOMPILE_DECODER_ONLY=on -DUSE_MKL=on -DCOMPILE_THREAD_VARIANT=on \
     -DUSE_WASM_COMPATIBLE_BLAS=off -DCOMPILE_MAIN_EIGEN=off
```
- [ ] There's a copy constructor deleted explicitly to not fire oneself in the foot added back by WASM for some reason. This can't be retained. Open to discussion.

File: app/marian-decoder-new.cpp
Comment (abhi-agg) [inline app/marian-decoder-new.cpp:1]: Now that you have this file as a main application for Service class, why do we need `app/main-mts.cpp` file? That file can be deleted. Right?

File: CMakeLists.txt
Comment (abhi-agg) [reply CMakeLists.txt:30]: COMPILE_WITH_PTHREADS is a marian specific cmake option. COMPILE_THREAD_VARIANT is specific to bergamot-translator project. We want to keep separate cmake options for both of them to control them individually.

File: app/marian-decoder-new.cpp
Comment (jerinphilip) [reply app/marian-decoder-new.cpp:1]: I'd rather keep it, it doesn't hurt anything.  Also:

> Integration failures are detected if same arguments work with service-cli and does not with bergamot-translator-app.

(doc/marian-integration.md)

Comment (jerinphilip) [reply app/marian-decoder-new.cpp:1]: The decoder is a different frontend to marian-decoder. This serves the purpose of comparing marian-decoder (upstream) with Service, for benchmarking. This is not the main application for Service.

File: CMakeLists.txt
Comment (kpu) [reply CMakeLists.txt:30]: Given that you said threads are on the roadmap, why are we spending so much time breaking threads?

**Comment by abhi-agg** [comment]
> changed the base branch from integration to wasm-integration 

@jerinphilip Why this switch?

**Comment by jerinphilip** [comment]
@abhi-agg: Just experimenting with GitHub interfaces and seeing if this made more sense. My adjustments (thread related) are for wasm, once resolved here you can go ahead with wasm-integration merging to integration. There are Batch and refactors, which crept in, but won't probably affect you. To some extent, these are already known to kpu. They're not much of a problem I think (I've tested my bits, so should work with wasm as well).

You adjust PCRE/PCRECPP this should be working with wasm. I've provided two ways for this depending how you want to retain wasm branch. 

Putting this on wasm-integration defers marian-dev diverging issues for later. And can be brought back when attempted to merge to integration/main.

File: src/translator/sentence_splitter.cpp
Comment (jerinphilip) [inline src/translator/sentence_splitter.cpp:33]: Convert this to `pcrecpp::StringPiece` with a bunch of ifdef guards and CMake, is an easy hack which will hurt in the long run. 

Ideal for both of us is a switch to PCRE2. @abhi-agg.

Comment (jerinphilip) [reply src/translator/sentence_splitter.cpp:33]: This place is the only bottleneck to merge this into wasm-integration (not integration), potentially. I've gotten builds succeeding till here, beyond this was too much work. I don't know an estiamte beyond this what errors will surface in builds. native is building, wasm fails with pcre2. 
@abhi-agg

Comment (jerinphilip) [reply src/translator/sentence_splitter.cpp:33]: [Patch](https://gist.github.com/jerinphilip/64679f42fa6929a3d2c860b16bb6f55e) which does the PCRECPP hack. Not going to push this in, would rather you switch to PCRE2. WASM [builds successfully](http://ix.io/2PZ2). At least on my development machine.

Comment (ugermann) [reply src/translator/sentence_splitter.cpp:33]: The problem with the way pcre / pcre2 is compiled here is that it does not seem to include ICU, which is what we need, because we need character properties for sentence splitting. I'd also rather not bundle pcre2 with ssplit-cpp, because in for native compilation one can use binary packages that are available for most distributions. In my oponion pcre2 with icu should be stand-alone projects so that they can be re-used elsewere. I've bee trying to compile icu with emscripten since the meeting yesterday without success.

Comment (jerinphilip) [reply src/translator/sentence_splitter.cpp:33]: @ugermann Linking https://github.com/ugermann/ssplit-cpp/issues/6 to move conversation there.

**Comment by abhi-agg** [comment]
> max-input-sentence-tokens -> max-length-break
> max-input-tokens -> mini-batch-words.

@jerinphilip Why this switch happened btw?

**Comment by kpu** [comment]
There was already an existing mini-batch-words option with the correct semantic.  It should have used that option.  

I think break just describes what it is doing better than sentence-tokens.

**Comment by jerinphilip** [comment]
`max-length-break` is named in accordance with `max-length-crop`, which truncates (or crops) the sentence. `-break` implying that it breaks sentences into smaller units of less than `max-length-break` if they end up larger after ssplit is being run.

**Comment by abhi-agg** [comment]
I confirm that wasm tests are running with the [updated ssplit-cpp submodule](https://github.com/ugermann/ssplit-cpp/pull/10).
=> Merging [this PR ](https://github.com/ugermann/ssplit-cpp/pull/10)will solve the issue.

File: wasm/test_page/bergamot.html
Comment (jerinphilip) [inline wasm/test_page/bergamot.html:41]: Are ? expected? I'm seeing something like this.
image

Comment (kpu) [reply wasm/test_page/bergamot.html:41]: Looks like a unicode disagreement.

**Comment by abhi-agg** [comment]
> From my side, the following are required.
> 
>     Mozilla should provide one easy switch to get pre-wasm, all targets compilation. My compilation looks like the following on this branch, which is not acceptable. Whatever CMakeFile.txt edits are present should be reviewed. I'm okay with existing marian, but I spent a lot of time trying to find these flags and turning them off to get compile to work. Also I expect the code I'm developing to be functioning with all of marian, not a select bunch of targets.
> 
> $ cmake .. -DCOMPILE_CUDA=off -DCMAKE_BUILD_TYPE=Release \
>      -DCOMPILE_DECODER_ONLY=on -DUSE_MKL=on -DCOMPILE_THREAD_VARIANT=on \
>      -DUSE_WASM_COMPATIBLE_BLAS=off -DCOMPILE_MAIN_EIGEN=off
> 
Based on the [latest CMakeLists.txt](https://github.com/browsermt/bergamot-translator/blob/jp/absorb-batch-translator/CMakeLists.txt#L18-L32)
- I don't know why do you have to set `-DCOMPILE_CUDA=off` because it is already `off` and it is supposed to be off for bergamot project
- `-DCMAKE_BUILD_TYPE=Release` is any way provided like this (we can also make the default build to release, I have no issues)
- `USE_MKL`
    - It should be off and will stay off for bergamot project because MKL is a proprietary library that can't be used if this project has to be used with Firefox (@kpu can confirm you this).
- `USE_WASM_COMPATIBLE_BLAS`
    - Is required for us right now because we are still waiting for solely intgemm dependent builds of marian from Nik (or whoever will do this). This has to stay until then. So, if you don't want to use it, you will have to keep it off in your builds. But they will stay `on` in this repo as default.
    - `-DCOMPILE_MAIN_EIGEN=off ` Why did you even use this switch? You have no concern at all with this. It is a switch that gets enabled when USE_WASM_COMPATIBLE_BLAS is set to ON. For your use case, you can set `-DUSE_WASM_COMPATIBLE_BLAS=off` and don't bother about this switch at all
- `COMPILE_THREAD_VARIANT`
    - Is a switch that enables/disabled multithreading code compilation in bergamot-translator source code. We have already discussed it why we need it for now. It is going to be removed in future 👍 


> There's a copy constructor deleted explicitly to not fire oneself in the foot added back by WASM for some reason. This can't be retained. Open to discussion.

As we discussed this before as well, I am completely aware of the implications. This code gets enabled only for wasm bindings. The bindings code will change and this will be removed from here later.

Let me know your final concerns :)

File: src/translator/sentence_splitter.cpp
Comment (abhi-agg) [reply src/translator/sentence_splitter.cpp:33]: Builds work with the merge of this [PR](https://github.com/ugermann/ssplit-cpp/pull/10)

**Comment by jerinphilip** [comment]
@abhi-agg I'm not satisfied, I'll move the concerns to issues (which will be brought up pre-merge into main). That said this isn't blocking merge to wasm-integration anymore? And it cleans out the threaded/non-threaded API removing the pcqueue.h modifications, fixes pcre2 (consequently sentencesplitting) and finalizes a bunch of options which are fixed before it propogates all over and we incur more debt fixing those.

**Comment by abhi-agg** [comment]
> I'm not satisfied, I'll move the concerns to issues (which will be brought up pre-merge into main).

@jerinphilip Could you please specify your concerns? I provided you a detailed explanation for your concerns regarding cmake switches. Let's resolve them here.

File: app/marian-decoder-new.cpp
Comment (abhi-agg) [inline app/marian-decoder-new.cpp:56]: @jerinphilip What does `marian_decoder_minimal` function do here?

File: app/marian-decoder-new.cpp
Comment (abhi-agg) [reply app/marian-decoder-new.cpp:1]: @jerinphilip Currently, there are 3 different main files in `app` folder of this repository. Technically, the app folder should contain only one that should demonstrate the use of unified APIs (which is done in `main.cpp`)

=> `main-mts.cpp` and this file will have to go away at sometime in future.

File: doc/marian-integration.md
Comment (abhi-agg) [inline doc/marian-integration.md:15]: Why are we compiling stuff with `USE_MKL=on`?
And why `USE_WASM_COMPATIBLE_BLAS` is set to `off`?

Comment (kpu) [reply doc/marian-integration.md:15]: You're not the only customer here.  If Jerin can make this efficient enough, it should replace the current queuing implementation in `marian-decoder`.  

In any case, we should always benchmark against the best possible native performance.

Comment (abhi-agg) [reply doc/marian-integration.md:15]: Okay. What is the goal of having this document? (PS: The instructions on how to compile bergamot-translator are already in top level README file).

Comment (abhi-agg) [reply doc/marian-integration.md]: Well. I fixed this. You have one cmake option that can help you compile vanilla marian.

**Review by abhi-agg — APPROVED** [review]
All concerns of @jerinphilip are dealt. So merging it.

**Pull request merged**
